ocrd_network/ocrd_network/deployer.py

```diff
@@ -0,0 +1,302 @@
+        client.close()
+        self.log.info('The MongoDB is stopped')
+
+    def kill_hosts(self) -> None:
+        self.log.debug('Starting to kill/stop hosts')
+        # Kill processing hosts
+        for host in self.hosts:
+            self.log.debug(f'Killing/Stopping processing workers on host: {host.config.address}')
+            if host.ssh_client:
+                host.ssh_client = create_ssh_client(host.config.address, host.config.username,
+                                                    host.config.password, host.config.keypath)
+            if host.docker_client:
+                host.docker_client = create_docker_client(host.config.address, host.config.username,
+                                                          host.config.password, host.config.keypath)
+            # Kill deployed OCR-D processor instances on this Processing worker host
+            self.kill_processing_worker(host)
+
+    def kill_processing_worker(self, host: HostData) -> None:
+        for pid in host.pids_native:
+            self.log.debug(f'Trying to kill/stop native processor: with PID: \'{pid}\'')
+            host.ssh_client.exec_command(f'kill {pid}')
+        host.pids_native = []
+
+        for pid in host.pids_docker:
+            self.log.debug(f'Trying to kill/stop docker container with PID: {pid}')
+            host.docker_client.containers.get(pid).stop()
+        host.pids_docker = []
+
+    def start_native_processor(self, client: SSHClient, processor_name: str, queue_url: str,
+                               database_url: str) -> str:
+        """ start a processor natively on a host via ssh
+
+        Args:
+            client:             paramiko SSHClient to execute commands on a host
+            processor_name:     name of processor to run
+            queue_url:          url to rabbitmq
+            database_url:       url to database
+
+        Returns:
+            str: pid of running process
+        """
+        # TODO: some processors are bashlib. They have to be started differently
+        self.log.info(f'Starting native processor: {processor_name}')
+        channel = client.invoke_shell()
+        stdin, stdout = channel.makefile('wb'), channel.makefile('rb')
+        cmd = f'{processor_name} --database {database_url} --queue {queue_url}'
+        # the only way (I could find) to make it work to start a process in the background and
+        # return early is this construction. The pid of the last started background process is
+        # printed with `echo $!` but it is printed inbetween other output. Because of that I added
+        # `xyz` before and after the code to easily be able to filter out the pid via regex when
+        # returning from the function
+        stdin.write(f'{cmd} & \n echo xyz$!xyz \n exit \n')
+        output = stdout.read().decode('utf-8')
+        stdout.close()
+        stdin.close()
+        return re_search(r'xyz([0-9]+)xyz', output).group(1)  # type: ignore
```

Comment:
I am surprised that the normal stdout and stderr is simply thrown away here. Shouldn't we try to persist this (perhaps in the job of the database)?

As to the problem of running asynchronously but being able to stop – now implemented by running in the bg, retrieving the PID, and killing: Would it perhaps be possible to implement this differently, by directly using Paramiko's channel object in the model?

(Roughly…
```python
transport = client.get_transport()
channel = transport.open_session()
channel.exec_command(cmd)
return channel.get_id()
```
…and then for `kill_processing_worker`…
```python
    chan = self.ssh_client.get_transport()._channels[id]
    chan.close()
```
…or perhaps with a dict of channels managed on our side.)

Reply:
> I am surprised that the normal stdout and stderr is simply thrown away here. Shouldn't we try to persist this (perhaps in the job of the database)?

IMO the worker stderr/stdout should at least be redirected to the server's own stderr/stdout (which in turn might be saved somewhere).

Reply:
I agree that the suggested solution is more elegant. However, I will let Jonas reply/resolve this and other related topics to deployment since he has a better understanding of how the ssh and docker clients work.

Reply:
Yes, the output should be stored that's definitely right. It is at least needed for debugging, I stumbled upon this myself. I am thinking of just logging it. This output here is just for the worker startup, the worker handles it's output itself. The worker should (and does so) log important information and do something if it has to be stopped unexpectedly. I think the output is "just" needed before the worker can get started and make the logging start.

Regarding the asynchronously running: This was the only way for me to make it work. I think for your suggestion to work the transport (or the channel, I don't know) has to be kept so it can be used later to shut it down. I currently prefer to start the workers and then let them run independently without an open connection or something like that. I think/thought (without a really good reason maybe) it is better to not keep a connection. Then other things would have to be worried about, keeping it going, handle failures etc.
But I have not much experience regarding this. What are the benefits of using the channel object and keeping it connected?

Reply:
> Yes, the output should be stored that's definitely right. It is at least needed for debugging, I stumbled upon this myself. I am thinking of just logging it. This output here is just for the worker startup, the worker handles it's output itself. The worker should (and does so) log important information and do something if it has to be stopped unexpectedly. I think the output is "just" needed before the worker can get started and make the logging start.

Understood. Yes, logging it (to `ocrd_network.processing_worker` logger as well) may be the best place IMO. (You'd have to re-wrap each line into a single log msg I believe, and ensure that stdout and stderr stay in-order, though...)

Reply:
> Regarding the asynchronously running: This was the only way for me to make it work. I think for your suggestion to work the transport (or the channel, I don't know) has to be kept so it can be used later to shut it down.

Yes.

> I currently prefer to start the workers and then let them run independently without an open connection or something like that. I think/thought (without a really good reason maybe) it is better to not keep a connection. Then other things would have to be worried about, keeping it going, handle failures etc.
> But I have not much experience regarding this. What are the benefits of using the channel object and keeping it connected?

Note: paramiko.transport by default does not send keep-alive messages, but can be [set up](https://docs.paramiko.org/en/stable/api/transport.html#paramiko.transport.Transport.set_keepalive) to do so.

IMO the benefit of keeping the session open would be that you don't have to worry about detaching and PID tracking in the first place. (The fact that paramiko does not provide access to the PID itself is a good argument against such tricks.)

But as I said in the other comment, perhaps this has to be seen in practice first, so let's try with your current implementation. 

(That being said, I do think there are better ways to attain the PID value. For example, you could do something like `echo $! > $processor.pid`, with `$processor` defined and passed by the client a priori, and then in `kill_processing_worker` do `kill $(cat $processor.pid)`. Or instead of `paramiko` use `asyncssh`, which directly provides a function [Connection.create_subprocess](https://asyncssh.readthedocs.io/en/latest/api.html#asyncssh.SSHClientConnection.create_subprocess) that lets you access its [get_pid](https://asyncssh.readthedocs.io/en/latest/api.html#asyncssh.SSHSubprocessTransport.get_pid). Or have the Processing Worker write its PID into the database under a new worker model. )

Reply:
I have now added at least a redirect for the output (stderr and stdout) to /tmp/ocrd-processing-server-startup.log, so that you can at least look somewhere for debugging.
The current solution works for the most part, but it is far from optimal, I definitely have to agree with Robert. Something should be changed here in the future. My first idea would be to start a tmux session for each processor. Then you could kill the processors more securely and you can also connect there for debugging and read the output there. However, I think we should discuss this question in one of our Monday meetings.

Reply:
continued in https://github.com/OCR-D/core/issues/1036